rust/eos-rs/src/api/monster.rs

```diff
@@ -0,0 +1,84 @@
+//! Structs and functions to interact with the data of monsters in a general context.
+
+use crate::api::objects::monster_catalog;
+use crate::ffi;
+use alloc::vec::Vec;
+
+/// Metadata of a monster species.
+///
+/// This struct provides general metadata about monster species in the game.
+pub struct MonsterSpeciesInfo(monster_catalog::Type);
+
+impl MonsterSpeciesInfo {
+    /// Returns the info struct for the monster species with the given ID.
+    ///
+    /// The caller should make sure, the ID is valid (refers to an existing monster species),
+    /// otherwise the data returned by methods of this struct will be invalid and undefined.
```

Comment:
This sounds like it should be `unsafe`. It's similar to things like `Vec::get_unchecked`.

Reply:
Right, I was thinking about this, but then again, most functions provided by the crate that call the game's functions with indices would be either unsafe and/or we'd need to check bounds checking behavior for all of them.

I'm on the fence about changing this one to be unsafe, because that would be inconsequential then. But I also don't want to go through all functions and check them... I think for this purpose leaving this valid with this note is acceptable. This is a special environment after all, and it's impossible to reliably follow Rust's rules in safe code here.

Reply:
I don't really feel strongly either way. I guess people should only ever be getting `monster_idx` from `monster_catalog` anyway.

Is there a better way to specify the type for `monster_idx`? Since `monster_catalog::Type` is just a type alias for `ctypes::c_uint`, I think the compiler won't be able to catch if you mistakenly pass in something like `item_catalog::ITEM_STICK` instead? I'm not sure how much work it'd be to define newtypes for enums instead.

Reply:
Yeah that was another thing I was considering, but there would be three options:
- Use enums instead (could even generate them via bindgen). Downside: You could not use any new Pokémon, since you would be hard-limited to the existing options.
- Use a newtype wrapper type. Downside is a bit more boilerplate in the code, but no runtime impact.
- Keep everything as is.

I think I prefer the second option. I would get rid of all the `*_catalog` re-exports in the `objects` module and come up with an easy to use newtype-based solution. Maybe even one that combines it with (somehow) auto-generated Enums based on the C-enums, so you can quickly construct them for existing Pokémon, items, etc.
I would make a new issue and do that after processing all the changes to pmdsky-debug if that sounds good.

Reply:
To visualize what I think (not tested to be valid code):
```rs
struct MonsterId(ffi::monster_id::Type);

impl MonsterId {
    pub unsafe fn new(id: ffi::monster_id::Type) {
        MonsterId(id)
    }

    pub fn bulbasaur() {
        MonsterId(ffi::monster_id::Bulbsaur)
    }
}
```

Another downside, is that the ffi-bindings would not use these wrapper types out of the box, but the raw enum C types. But maybe there's also a solution for that.

Reply:
Okay, just checked how Bindgen does it, if you specify it as a NewType. I think this is pretty good?:
```rs
/* ... */
impl monster_id {
    pub const MONSTER_RESERVE_42: monster_id = monster_id(596);
}
impl monster_id {
    pub const MONSTER_RESERVE_43: monster_id = monster_id(597);
}
impl monster_id {
    pub const MONSTER_RESERVE_44: monster_id = monster_id(598);
}
impl monster_id {
    pub const MONSTER_RESERVE_45: monster_id = monster_id(599);
}
#[repr(transparent)]
#[derive(Clone, Hash, PartialEq, Eq)]
pub struct monster_id(pub crate::ctypes::c_uint);
```

In the high-level code I would make a type alias an provide extra impls, including moving the ones from `api::monster` to this newtype.

```
pub type MonsterId = ffi::monster_id;

impl MonsterId {
    /// Returns the ID of this monster.
    pub const fn id(&self) -> crate::ctypes::c_uint {
        self.0
    }

    /// Returns the gender field of the monster.
    pub fn gender(&self) -> u8 {
        // TODO: Enum.
        unsafe { ffi::GetMonsterGender(self.0) }
    }
    /* ... */
}
```

Reply:
This seems pretty reasonable to me.

Reply:
How about making the `id` function unsafe in this example? Users could safely use the existing Pokémon as an index, but if they want to create a new `MonsterSpecies`, they need to ensure that valid data exists for it

Reply:
Yeah, that would be a new or get method though, since id just returns an existing ID. Only downside is that bindgen generates the newtypes with public fields, so you could manually construct it, but eh.